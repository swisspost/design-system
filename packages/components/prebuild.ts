import fs from 'fs';
import path from 'path';
import { format } from 'prettier';

interface ComponentNameOutputOptions {
  type: 'json' | 'scss';
  template: string;
  lineSeparator: string;
  // https://prettier.io/docs/options#parser
  // if formatParser is null, the parser will be derived from the "type" field above
  formatParser: 'json' | 'scss' | null;
}

const DEFAULT_OUTPUT_OPTIONS: IComponentNameOutputOptions = {
  type: 'json',
  template: '{{names}}',
  lineSeparator: ',',
  formatParser: null,
};

export const componentNameOutputOptions: Partial<IComponentNameOutputOptions>[] = [
  {
    template: '{ componentNames: [{{names}}] }',
  },
  {
    type: 'scss',
    template: '$component-names: ({{names}});',
  },
];

export async function createComponentNameOutput(
  outputOptionsArray: Partial<IComponentNameOutputOptions>[],
) {
  // Define the source and output paths
  const SOURCE_PATH: string[] = [path.resolve('src/components')];
  const OUTPUT_PATH: string = path.resolve('src/_generated');

  // Process component names
  const componentNames: string[] = SOURCE_PATH.reduce((names: string[], srcPath: string) => {
    if (fs.existsSync(srcPath)) {
      const fileNames: string[] = fs.readdirSync(srcPath);
      const camelCaseNames: string[] = fileNames.map((name: string) =>
        name.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase(),
      );
      names.push(...camelCaseNames);
    }
    return names;
  }, []);

  // Ensure the output directory exists
  fs.mkdirSync(OUTPUT_PATH, { recursive: true });

  await Promise.all(
    outputOptionsArray.map(async (outputOptions = {}) => {
      const options: IComponentNameOutputOptions = { ...DEFAULT_OUTPUT_OPTIONS, ...outputOptions };
      const names = componentNames.map(n => `'${n}'`).join(options.lineSeparator);
      const template = options.template.replace('{{names}}', names);
      const output = await format(template, { parser: options.formatParser || options.type });

      // Write the SCSS file
      fs.writeFileSync(`${OUTPUT_PATH}/component-names.${options.type}`, output, 'utf8');
    }),
  );
}

function prebuild() {
  createComponentNameOutput(componentNameOutputOptions);
}

prebuild();
