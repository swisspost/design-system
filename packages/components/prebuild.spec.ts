import fs from 'fs';
import mockFs from 'mock-fs';
import { createComponentNamesScssMap } from './prebuild';

describe('prebuild', () => {
  it('should read out component folders', () => {
    mockFs({
      'src/components': {
        'post-accordion': {},
        'post-avatar': {},
        'post-back-to-tlop': {},
        'post-breadcrumbs': {},
        'post-breadcrumbs-item': {},
        'post-123-component': {},
        'some-other-folder': {},
        'folder_name_with_different_notation': {},
      },
      'src/styles/generated': {},
    });

    createComponentNamesScssMap();
    const componentNamesScssMap = fs.readFileSync(
      'src/styles/generated/component-names.scss',
      'utf8',
    );

    mockFs.restore();

    expect(componentNamesScssMap).toMatchSnapshot();
  });
});
