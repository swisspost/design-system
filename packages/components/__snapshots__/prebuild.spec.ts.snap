// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prebuild should read out component folders 1`] = `
"$component-names: (
  folder_name_with_different_notation,
  post-123-component,
  post-accordion,
  post-avatar,
  post-back-to-tlop,
  post-breadcrumbs,
  post-breadcrumbs-item,
  some-other-folder
);
"
`;
